lib/kinesis_client.rb

```diff
@@ -99,8 +93,23 @@ def push_records
   end
 
   def filter_failures(resp)
-    resp.records.filter_map.with_index do |record, i|
-      avro.decode(@records[i + @batch_size * @batch_count]) if record.responds_to?(:error_message)
+    failed_records = resp.records.filter_map.with_index do |record, i|
+      
+      @records[i + @batch_size * @batch_count] if record.responds_to?(:error_message)
```

Comment:
I'm a little nervous about `@batch_count` getting out of sync with the other variables. For example, when you modify `@records` in the `retry_failed_records` method, should `@batch_count` be updated?

Reply:
I think the way to ensure that this works as intended is to set `@batch_count = 0` at the start of `push_records`. I don't think it has any relevance outside of that context, right?

It even could just be a local variable there that gets passed in there, I don't think anything needs to interact with it from outside that method, it's just tracking our position within the current set of records. (or even just `each_slice.with_index` would be more foolproof?)